README.md

```diff
@@ -1,16 +1,659 @@
+| `prf` | `[&UCAN]`                                    |           | The CIDs of the UCANs that provide the authority to run | Yes      |         |
+| `run` | `"*" or {String: {URI : {Ability : [Any]}}}` |           | Which UCAN capabilities to run                          | Yes      |         |
+| `nnc` | `String`                                     |           | A unique nonce, to distinguish each invocation          | Yes      |         |
+| `ext` | `Any`                                        |           | Non-normative extended fields                           | No       | `null`  |
+| `sig` | `Bytes`                                      |           | Signature of the rest of the field canonicalized        | Yes      |         |
+  
+### 3.1.1 Proofs
+
+The `prf` field MUST contain CIDs pointing to the UCANs that provide the authority to run these actions. Restricting the outmost UCANs to only the authority required for the current invocation is RECOMMENDED.
+
+### 3.1.2 Version
+
+The `v` field MUST contain the version of the invocation object  schema.
+
+### 3.1.3 Run Capabilities
+
+The `run` field MUST reference the actions contained in the UCAN are to be run during the invocation. To run all actions in the underlying UCAN, the `"*"` value MUST be used. If only specific actions (or [pipelines](#5-promise-pipelining)) are intended to be run, then they MUST be prefixed with an arbitrary label and treated as a UCAN attenuation: all actions MUST be backed by a matching capability of equal or greater authority.
+
+#### 3.1.3.1 Promises
+
+The only difference from general capabilities is that [promises](#5-promise-pipelining) MAY also be used as inputs to attenuated fields.
+
+If a capability input has the key `"_"` and the value is a promise, the input MUST be discarded and only used for determining sequencing actions.
+
+### 3.1.4 Nonce
+
+The `nnc` field MUST contain a unique nonce. This field exists to enable making the CID of each invocation unique. While this field MAY be an empty string, it is NOT RECOMMENDED.
+
+### 3.1.5 Extended Fields
+
+The OPTIONAL `ext` field MAY contain arbitrary data. If not present, the `ext` field MUST be interpreted as `null`, including for [signature](#315-signature).
+
+### 3.1.6 Signature
+
+The `sig` field MUST contain the signature of the other fields. To construct the payload, the other fields MUST first be canonicalized as `dag-cbor`.
+
+If serialized as JSON, the `sig` field MUST be serialized as [unpadded base64url](https://datatracker.ietf.org/doc/html/rfc4648#section-5).
+
+## 3.2 IPLD Schema
+
+``` ipldsch
+type Invocation struct {
+  prf [&UCAN] -- The UCANs providing authority
+  v   SemVer  -- Version
+  run Scope   -- Which actions to invoke
+  nnc String  -- Nonce
+  ext nullable Any (implicit null) -- Extended fields
+}
+
+type SignedInvocation struct {
+  inv Invocation (rename "ucan/invoke") 
+  sig VarSig  -- Signature
+}
```

Comment:
```suggestion
type SignedInvocation struct {
  inv &Invocation (rename "ucan/invoke") 
  sig VarSig  -- Signature
}
```

Making the `inv` property a CID would make canonicalization extremely straight forward. You would just sign over the CID bytes.

Reply:
Hmm yeah that's a good point 🤔 Doesn't DAG-* already get us canonicalization?

I wonder if Varsig will help with this too, if we include this: https://github.com/ChainAgnostic/varsig/issues/5